File: exams/u3/sigcha/Exam 3/SortApplication/sort_app/model/quick_sort.py
from typing import List
from sort_app.model.sorting_strategy import SortingStrategy

class QuickSort(SortingStrategy):
    def sort(self, data: List[int]) -> List[int]:
        sorted_list = data[:]
        self._quick_sort(sorted_list, 0, len(sorted_list) - 1)
        return sorted_list

    def _quick_sort(self, array: List[int], low: int, high: int):
        if low < high:
            pi = self._partition(array, low, high)
            self._quick_sort(array, low, pi - 1)
            self._quick_sort(array, pi + 1, high)

    def _partition(self, array: List[int], low: int, high: int) -> int:
        pivot = array[high]
        i = low - 1
        for j in range(low, high):
            if array[j] < pivot:
                i += 1
                array[i], array[j] = array[j], array[i]
        array[i + 1], array[high] = array[high], array[i + 1]
        return i + 1

File: exams/u3/toscano/Evaluacion_Damian_Toscano/Evaluacion_Toscano/controller/QuickSort.js
const SortingStrategy = require('./SortingStrategy');

class QuickSort extends SortingStrategy {
    sort(data) {
        console.log("Quick sort");
        return this.quickSort(data, 0, data.length - 1);
    }

    quickSort(arr, low, high) {
        if (low < high) {
            let pi = this.partition(arr, low, high);
            this.quickSort(arr, low, pi - 1);
            this.quickSort(arr, pi + 1, high);
        }
        return arr;
    }

    partition(arr, low, high) {
        let pivot = arr[high];
        let i = (low - 1);
        for (let j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
        }
        [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
        return i + 1;
    }
}

module.exports = QuickSort;
